File: skills/adk-agentic-prod-workflows/assets/workflow-templates/ts/agent-server/tools.ts
/**
 * Tool definitions for ADK Agent Server.
 * Each tool has a clear signature, parameter schema, and error handling.
 */

import type { ToolContext } from "@google/adk";

/**
 * Search internal knowledge base.
 * Production: replace with actual vector DB or RAG pipeline.
 */
export async function searchKnowledgeBase(
  query: string,
  topK: number = 5,
  ctx?: ToolContext
): Promise<{ results: Array<{ title: string; content: string; score: number }> }> {
  // Parameterized — never interpolate user input directly
  const sanitizedQuery = query.trim().slice(0, 500);

  // Placeholder — replace with actual search
  return {
    results: [
      {
        title: `Result for: ${sanitizedQuery}`,
        content: "Relevant information found in knowledge base.",
        score: 0.92,
      },
    ],
  };
}

/**
 * Fetch data from external API.
 * Production: replace with actual HTTP client call.
 */
export async function fetchExternalData(
  endpoint: string,
  params?: Record<string, string>,
  ctx?: ToolContext
): Promise<{ status: number; data: unknown }> {
  // Validate endpoint
  const allowedEndpoints = ["weather", "news", "stocks"];
  const base = endpoint.split("/")[0];
  if (!allowedEndpoints.includes(base)) {
    throw new Error(`Endpoint '${base}' not allowed. Allowed: ${allowedEndpoints.join(", ")}`);
  }

  // Placeholder — replace with actual HTTP call
  return {
    status: 200,
    data: { endpoint, params, result: "Data fetched successfully" },
  };
}

/**
 * Quality gate tool — call when output meets requirements.
 * Escalates the event to signal loop completion.
 */
export async function qualityGate(
  score: number,
  threshold: number = 0.8,
  ctx?: ToolContext
): Promise<{ passed: boolean; score: number; threshold: number }> {
  const passed = score >= threshold;
  if (passed && ctx) {
    ctx.actions.escalate = true;
  }
  return { passed, score, threshold };
}

File: .claude/skills/adk-agentic-prod-workflows/assets/workflow-templates/python/template-loop/tools.py
"""Loop workflow: tool definitions."""

from google.adk.tools.tool_context import ToolContext


def build_exit_loop_tool():
    """Quality gate tool — call when output meets requirements to exit the loop."""

    def exit_loop(tool_context: ToolContext):
        """Signal that the current output meets quality requirements.

        Call this ONLY when the content is good enough to exit the refinement loop.
        This escalates the event, causing LoopAgent to stop iterating.
        """
        tool_context.actions.escalate = True
        return {"status": "quality_gate_passed"}

    return exit_loop


def build_quality_score_tool(min_score: float = 0.8):
    """Tool that checks a quality score against a threshold."""

    def check_quality(score: float, tool_context: ToolContext):
        """Check if quality score meets the threshold. Escalates if passed.

        Args:
            score: Quality score between 0.0 and 1.0.
        """
        if score >= min_score:
            tool_context.actions.escalate = True
            return {"passed": True, "score": score, "threshold": min_score}
        return {"passed": False, "score": score, "threshold": min_score}

    return check_quality
